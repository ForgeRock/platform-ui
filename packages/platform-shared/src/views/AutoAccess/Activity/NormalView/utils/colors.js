/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

export const colorsHex = ['#2ed47a', '#ffb946', '#109cf1', '#f7685b', '#61e2e2'];
export const colorsReadable = ['green', 'yellow', 'blue', 'red', 'cyan'];
