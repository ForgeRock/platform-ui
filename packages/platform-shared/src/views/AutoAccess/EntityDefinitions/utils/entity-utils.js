/**
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { v4 as uuidv4 } from 'uuid';

import materialIconNames from './icon-options';

export const defaultEntity = {
  id: uuidv4(),
  isDirty: false,
  description: '',
  schemas: {
    _meta: {
      type: 'managed',
      indexing: {
        latest: 'false',
        log: 'true',
      },
      icon: '',
      displayName: '',
      properties: {},
    },
    properties: {},
  },
};

export const defaultAttribute = {
  id: uuidv4(),
  isInternal: false,
  isPrimaryKey: false,
  isRequired: false,
  isAutoGenerate: false,
  ml: {
    isFeature: false,
    isLabel: false,
  },
  display: {
    name: '',
    group: 'General',
    order: 10,
    isVisible: false,
    isReadOnly: false,
    isSearchable: false,
  },
};

export const getPrimaryKeyIfExistsOrFirstColumn = (entity) => {
  let primaryKey = null;

  Object.keys(entity.schemas._meta.properties ? entity.schemas._meta.properties : {}).forEach((attributeName) => {
    if (entity.schemas._meta.properties) {
      const attribute = entity.schemas._meta.properties[attributeName];
      if (attribute.isPrimaryKey && !primaryKey) {
        primaryKey = attributeName;
      }
    }
  });

  if (!primaryKey) {
    primaryKey = 'metadata.primaryKey';
  }

  return primaryKey;
};

export const isPrimaryKeyPresent = (entity) => {
  let isPresent = false;

  Object.keys(entity.schemas._meta.properties ? entity.schemas._meta.properties : {}).forEach((attributeName) => {
    if (entity.schemas._meta.properties[attributeName].isPrimaryKey) {
      isPresent = true;
    }
  });

  return isPresent;
};

export const getPrimaryKey = (entity) => {
  let primaryKey = null;

  Object.keys(entity.schemas._meta.properties ? entity.schemas._meta.properties : {}).forEach((attributeName) => {
    if (entity.schemas._meta.properties) {
      const attribute = entity.schemas._meta.properties[attributeName];
      if (attribute.isPrimaryKey && !primaryKey) {
        primaryKey = attributeName;
      }
    }
  });

  return primaryKey;
};

export const cleanUpEntityDefinition = (entity) => {
  if (entity) {
    entity.id = uuidv4();

    if (!entity.schemas) {
      entity.schemas = { ...defaultEntity.schemas };
    }
    if (entity?.schemas?.properties && !entity?.schemas?._meta?.properties) {
      entity.schemas._meta = {
        type: 'managed',
        indexing: {
          latest: 'false',
          log: 'true',
        },
        icon: '',
        displayName: '',
        properties: {},
      };
      Object.keys(entity.schemas.properties).forEach((attributeName) => {
        entity.schemas._meta.properties[attributeName] = { ...defaultAttribute };
      });
    }
    if (!entity.schemas._meta) {
      entity.schemas._meta = { ...defaultEntity.schemas._meta };
    }
    if (!entity.schemas._meta.displayName) {
      entity.schemas._meta.displayName = `${entity.namespace}/${entity.name}`;
    }

    if (!entity.schemas._meta.icon) {
      entity.schemas._meta.icon = 'apps';
    } else if (!materialIconNames.includes(entity.schemas._meta.icon)) {
      entity.schemas._meta.icon = 'apps';
    }

    if (!entity.schemas._meta.properties) {
      entity.schemas._meta.properties = {};
    }

    if (!entity.schemas.properties) {
      entity.schemas.properties = {};
    }

    Object.keys(entity.schemas._meta.properties).forEach((attributeName) => {
      const attribute = entity.schemas._meta.properties[attributeName];
      attribute.display = { ...defaultAttribute.display, ...attribute.display };
      attribute.ml = { ...defaultAttribute.ml, ...attribute.ml };
    });
  }
  return entity;
};

export const cleanUpEntity = (newEntity) => {
  const myEntity = { ...newEntity };
  delete myEntity.id;
  delete myEntity.isDirty;
  delete myEntity.created;
  delete myEntity.name;
  delete myEntity.namespace;
  delete myEntity.tenantId;
  delete myEntity.version;
  delete myEntity.open$;
  delete myEntity.index$;
  delete myEntity.createdMillis;
  delete myEntity.loaded;
  return myEntity;
};
