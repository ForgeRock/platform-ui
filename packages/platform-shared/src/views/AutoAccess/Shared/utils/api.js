/**
 * Copyright (c) 2022-2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import _ from 'lodash';
import { generateAutoAccessJas } from '@forgerock/platform-shared/src/api/BaseApi';
import store from '@/store';

/**
 * Format an object to be used as the payload for an elastic query search
 *
 * @param {Array} dateRange - Start and end date of the query
 * @param {Object} filterObject - Object containing risk range, geo coords, and filters
 * @param {String} userId - User id for specific user query
 * @returns {Object} Elastic query search compatible object
 */
export function getQueryFilters(dateRange, filterObject, userId) {
  // Date Range
  const bool = {
    must: [{
      range: {
        'predictionResult.features.timestamp': {
          gte: dateRange.startDate,
          lte: dateRange.endDate,
        },
      },
    },
    ],
  };

  // Risk Range
  if (filterObject.riskRange) {
    bool.must.push({
      range: {
        'predictionResult.risk_score_data.risk_score': {
          lt: filterObject.riskRange[1] + 0.5,
          gte: filterObject.riskRange[0],
        },
      },
    });
  }

  // Geo Coordinates
  if (Object.keys(filterObject.geoCoordinates).length) {
    bool.must.push({
      range: {
        'predictionResult.features.geoData.lat': {
          lte: filterObject.geoCoordinates.north,
          gte: filterObject.geoCoordinates.south,
        },
      },
    });
    bool.must.push({
      range: {
        'predictionResult.features.geoData.longitude': {
          lte: filterObject.geoCoordinates.east,
          gte: filterObject.geoCoordinates.west,
        },
      },
    });
  }

  // Is Risky Event
  if (filterObject.is_risky_event) {
    bool.must.push({
      term: {
        'predictionResult.risk_score_data.is_risky_event': filterObject.is_risky_event,
      },
    });
  }

  // Attribute Filters
  const featureMap = {};
  filterObject.features.forEach((feature) => {
    if (!featureMap[feature.key]) {
      featureMap[feature.key] = [];
    }
    featureMap[feature.key].push(...feature.value);
  });

  Object.keys(featureMap).forEach((key) => {
    bool.must.push({
      bool: {
        should: featureMap[key].map((value) => (
          {
            term: {
              [`${key}.keyword`]: value,
            },
          }
        )),
        minimum_should_match: 1,
      },
    });
  });

  bool.must.push(
    {
      term: {
        realm: `/${store.state.realm}`,
      },
    },
  );

  // User
  if (userId) {
    bool.must.push({
      term: {
        'predictionResult.features.userId': this.userId,
      },
    });
  }

  const clusteringReasons = [];
  const heuristicRawResultsReasons = [];
  let isIpBlocked = false;
  let isAdvancedBotDetection = false;
  const uebaReasons = [];
  filterObject.reasons.forEach((reasons) => {
    reasons.forEach((reason) => {
      if (reason === 'is_ip_blocked') {
        isIpBlocked = true;
      } else if (reason === 'is_advanced_bot_detection') {
        isAdvancedBotDetection = true;
      } else if (reason.indexOf('is_') === 0) {
        heuristicRawResultsReasons.push(reason);
      }
      if (store.state.Dashboard.uebaReasons.includes(reason)) {
        uebaReasons.push(reason);
      }
      if (store.state.Dashboard.clusteringReasons.includes(reason)) {
        clusteringReasons.push(reason);
      }
    });
  });

  // Heuristic, UEBA, Clustering Reason
  const allQuery = { bool: { should: [], minimum_should_match: 1 } };
  if (filterObject.reasons.length > 0) {
    const hueristicsRawResultsQuery = [];

    // Heuristics IP Block Rule
    if (isIpBlocked) {
      allQuery.bool.should.push({
        bool: {
          should: {
            term: {
              'predictionResult.risk_score_data.heuristic_agg_result.block_rule_result.is_blocked': true,
            },
          },
          minimum_should_match: 1,
        },
      });
    }

    // Heuristics Advanced Bot Detection
    if (isAdvancedBotDetection) {
      allQuery.bool.should.push({
        bool: {
          should: {
            term: {
              'predictionResult.botD.isBot': true,
            },
          },
          minimum_should_match: 1,
        },
      });
    }

    // Heuristic Raw Results
    if (heuristicRawResultsReasons.length > 0) {
      hueristicsRawResultsQuery.push({
        bool: {
          should: heuristicRawResultsReasons.map((heuristic) => ({
            term: {
              [`predictionResult.risk_score_data.heuristic_agg_result.raw_results.${heuristic}`]: true,
            },
          })),
          minimum_should_match: 1,
        },
      });

      allQuery.bool.should.push({
        nested: {
          path: 'predictionResult.risk_score_data.heuristic_agg_result.raw_results',
          query: {
            bool: {
              should: hueristicsRawResultsQuery,
            },
          },
        },
      });
    }

    // UEBA
    if (uebaReasons.length > 0) {
      const uebaQuery = uebaReasons.map((reason) => ({
        term: {
          'predictionResult.ueba_signal.explainable_features': reason,
        },
      }));

      allQuery.bool.should.push({
        bool: {
          filter: {
            script: {
              script: "return doc['predictionResult.risk_score_data.ueba_avg_risk_score'].value > doc['predictionResult.risk_score_data.risk_score_threshold'].value;",
            },
          },
          should: uebaQuery,
          minimum_should_match: 1,
        },
      });
    }

    // Clustering
    if (clusteringReasons.length > 0) {
      const clusteringQuery = clusteringReasons.map((reason) => ({
        term: {
          'predictionResult.risk_score_data.clustering_result.top_cluster_explainability': reason,
        },
      }));

      allQuery.bool.should.push({
        bool: {
          filter: {
            script: {
              script: "return doc['predictionResult.risk_score_data.clustering_model_risk_score'].value > doc['predictionResult.risk_score_data.risk_score_threshold'].value;",
            },
          },
          should: clusteringQuery,
          minimum_should_match: 1,
        },
      });
    }

    bool.must.push(allQuery);
  }

  return bool;
}

export const getConfig = () => new Promise((resolve, reject) => {
  const param = {
    indexName: 'risk_explainability',
    query: {
      query: {
        term: {
          'predictionResult.risk_score_data.is_risky_event': true,
        },
      },
      size: 1,
    },
  };
  generateAutoAccessJas().post('/entity/search/', param)
    .then((data) => {
      resolve({
        thresholds: {
          high: _.get(data, 'data.hits.hits[0]._source.predictionResult.risk_score_data.risk_score_threshold', 0),
        },
      });
    })
    .catch((e) => {
      reject(e);
    });
});
