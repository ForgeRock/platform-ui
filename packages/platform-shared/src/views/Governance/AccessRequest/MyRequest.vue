<!-- Copyright (c) 2023-2024 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div>
    <BContainer>
      <!-- Header -->
      <BRow class="mt-5 pb-4 pb-lg-0 align-items-center">
        <BCol lg="8">
          <FrHeader
            :title="$t('pageTitles.MyRequests')"
            :subtitle="$t('pages.myRequests.subTitle')" />
        </BCol>
        <BCol
          v-if="!userStore.adminUser"
          lg="4">
          <div class="d-flex justify-content-lg-end">
            <BButton
              variant="primary"
              :aria-label="$t('governance.accessRequest.newRequest.newRequest')"
              @click="newRequest">
              <FrIcon
                icon-class="mr-2"
                name="add">
                {{ $t('governance.accessRequest.newRequest.newRequest') }}
              </FrIcon>
            </BButton>
          </div>
        </BCol>
      </BRow>
      <!-- Content -->
      <BRow>
        <BCol>
          <BCard no-body>
            <FrAccessRequestList
              :is-loading="isLoading"
              :requests="accessRequests"
              @open-detail="viewDetails">
              <template #header>
                <FrRequestToolbar
                  :status-options="statusOptions"
                  @filter-change="filterHandler({ filter: $event })"
                  @sort-change="filterHandler({ sortKeys: $event })"
                  @sort-direction-change="filterHandler({ sortDir: $event })"
                  @status-change="filterHandler({ status: $event })" />
              </template>
              <template #no-data>
                <FrNoData
                  class="mb-4 border-top"
                  data-testid="requests-no-data"
                  icon="person_add"
                  :card="false"
                  :subtitle="$t('governance.accessRequest.noRequests', { status: getStatusText(statusOptions, status) })" />
              </template>
              <template #actions="{ item }">
                <div class="d-flex justify-content-end align-items-center">
                  <div class="d-none d-lg-block mr-4">
                    <BBadge
                      class="w-100px font-weight-normal"
                      data-testid="status-badge"
                      :variant="status === 'complete' ? 'success' : 'light'">
                      {{ getStatusText(statusOptions, status) }}
                    </BBadge>
                  </div>
                  <FrActionsCell
                    test-id="cell-button"
                    :delete-option="false"
                    :divider="false"
                    :edit-option="false">
                    <template #custom-top-actions>
                      <BDropdownItem
                        data-testid="view-details-button"
                        @click="viewDetails(item)">
                        <FrIcon
                          icon-class="mr-2"
                          name="list_alt">
                          {{ $t('common.viewDetails') }}
                        </FrIcon>
                      </BDropdownItem>
                      <template v-if="status === 'in-progress'">
                        <BDropdownDivider />
                        <BDropdownItem
                          v-if="userStore.adminUser"
                          @click="openModal(item, 'REASSIGN')">
                          <FrIcon
                            icon-class="mr-2"
                            name="redo">
                            {{ $t('common.forward') }}
                          </FrIcon>
                        </Bdropdownitem>
                        <BDropdownItem @click="openModal(item, 'CANCEL')">
                          <FrIcon
                            icon-class="mr-2"
                            name="cancel">
                            {{ $t('governance.accessRequest.myRequests.cancelRequest') }}
                          </FrIcon>
                        </Bdropdownitem>
                      </template>
                    </template>
                  </FrActionsCell>
                </div>
              </template>
            </FrAccessRequestList>
            <FrPagination
              v-if="totalRows > 0"
              v-model="currentPage"
              :per-page="pageSize"
              :total-rows="totalRows"
              @input="filterHandler({ currentPage: $event })"
              @on-page-size-change="filterHandler({ pageSize: $event })" />
          </BCard>
        </BCol>
      </BRow>
    </BContainer>
    <FrNewRequestModal />
    <FrRequestModal
      :type="modalType"
      :item="modalItem"
      @modal-closed="modalItem = {}"
      @update-list="loadRequests(true)" />
  </div>
</template>

<script setup>
/**
 * Displays the list of Requests and the corresponding filters.
 * @component MyRequest
 */
import {
  BBadge,
  BButton,
  BCard,
  BCol,
  BContainer,
  BDropdownDivider,
  BDropdownItem,
  BRow,
} from 'bootstrap-vue';
import {
  computed,
  onMounted,
  ref,
} from 'vue';
import { useUserStore } from '@forgerock/platform-shared/src/stores/user';
import useBvModal from '@forgerock/platform-shared/src/composables/bvModal';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
import FrPagination from '@forgerock/platform-shared/src/components/Pagination';
import FrActionsCell from '@forgerock/platform-shared/src/components/cells/ActionsCell';
import FrNoData from '@forgerock/platform-shared/src/components/NoData';
import {
  getRequestFilter,
  getStatusText,
  sortKeysMap,
} from '@forgerock/platform-shared/src/utils/governance/AccessRequestUtils';
import { REQUEST_MODAL_TYPES } from '@forgerock/platform-shared/src/utils/governance/constants';
import { getRequests, getUserRequests } from '@forgerock/platform-shared/src/api/governance/AccessRequestApi';
import FrAccessRequestList from '@forgerock/platform-shared/src/components/governance/AccessRequestList';
import FrRequestToolbar from '@forgerock/platform-shared/src/components/governance/RequestToolbar';
import FrNewRequestModal from '@forgerock/platform-shared/src/components/governance/NewRequestModal';
import FrRequestModal from '@forgerock/platform-shared/src/components/governance/RequestModal/RequestModal';
import { showErrorMessage } from '@forgerock/platform-shared/src/utils/notification';
import { useRouter } from 'vue-router';
import i18n from '@/i18n';

// Composables
const { userId, adminUser } = useUserStore();
const { bvModal } = useBvModal();
const router = useRouter();

const userStore = computed(() => ({
  userId,
  adminUser,
}));

const accessRequests = ref([]);
const currentPage = ref(1);
const filter = ref({});
const isLoading = ref(false);
const modalItem = ref({});
const modalType = ref(REQUEST_MODAL_TYPES.CANCEL);
const pageSize = ref(10);
const sortDir = ref('desc');
const sortKeys = ref('date');
const status = ref('in-progress');
const statusOptions = ref([
  {
    text: i18n.global.t('governance.status.pending'),
    value: 'in-progress',
  },
  {
    text: i18n.global.t('governance.status.complete'),
    value: 'complete',
  },
  {
    text: i18n.global.t('governance.status.canceled'),
    value: 'cancelled',
  },
]);
const totalRows = ref(0);

const componentRefs = new Map([
  ['currentPage', currentPage],
  ['filter', filter],
  ['pageSize', pageSize],
  ['sortDir', sortDir],
  ['sortDir', sortDir],
  ['sortKeys', sortKeys],
  ['status', status],
  ['totalRows', totalRows],
]);

/**
 * Get current users access requests based on query params and target filter
 */
async function loadRequests(goToFirstPage) {
  isLoading.value = true;

  if (goToFirstPage) currentPage.value = 1;
  const payload = getRequestFilter(filter.value, status.value);
  const params = {
    pagedResultsOffset: (currentPage.value - 1) * pageSize.value,
    pageSize: pageSize.value,
    sortKeys: sortKeysMap[sortKeys.value],
    sortDir: sortDir.value,
  };

  if (sortKeys.value === 'date') params.sortType = 'date';

  try {
    const { data } = userStore.value.adminUser
      ? await getRequests(params, payload)
      : await getUserRequests(userId, params, payload);
    accessRequests.value = data.result;
    totalRows.value = data.totalCount;
  } catch (error) {
    accessRequests.value = [];
    totalRows.value = 0;
    showErrorMessage(error, i18n.global.t('governance.accessRequest.myRequests.errorGettingRequests'));
  } finally {
    isLoading.value = false;
  }
}

function newRequest() {
  bvModal.value.show('NewRequestModal');
}

function viewDetails(item) {
  router.push({
    name: 'MyRequestDetails',
    params: { requestId: item.details.id },
  });
}

function openModal(item, type) {
  modalItem.value = item;
  modalType.value = REQUEST_MODAL_TYPES[type];
  bvModal.value.show('request_modal');
}

/**
 * Handles filtering requests as well as updates to pagination
 * @param {Object} property updated property
 */
function filterHandler(property) {
  const [key, value] = Object.entries(property)[0];
  componentRefs.get(key).value = value;
  const resetPaging = (key !== 'currentPage');
  loadRequests(resetPaging);
}

onMounted(async () => {
  loadRequests();
});
</script>
