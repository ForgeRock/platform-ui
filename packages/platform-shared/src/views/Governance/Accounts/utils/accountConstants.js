/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

const ACCOUNT_TYPES = {
  DEFAULT: 'default',
  MACHINE: 'machine',
};

const ACCOUNT_SUBTYPES = [
  {
    value: 'oauthToken', text: 'OAuth Token', icon: 'key', iconColor: 'blue',
  },
  {
    value: 'serviceAccount', text: 'Service Account', icon: 'badge', iconColor: 'purple',
  },
  {
    value: 'rpa', text: 'RPA', icon: 'precision_manufacturing', iconColor: 'orange',
  },
  {
    value: 'bot', text: 'BOT', icon: 'smart_toy', iconColor: 'green',
  },
  {
    value: 'accessToken', text: 'Access Token', icon: 'vpn_key', iconColor: 'brown',
  },
  {
    value: 'secret', text: 'Secret', icon: 'lock', iconColor: 'red',
  },
  {
    value: 'apiKey', text: 'API Key', icon: 'code', iconColor: 'blue',
  },
  {
    value: 'personalAccessToken', text: 'Personal Access Token', icon: 'person', iconColor: 'orange',
  },
  {
    value: 'managedAccount', text: 'Managed Account', icon: 'manage_accounts', iconColor: 'blue',
  },
  {
    value: 'teamToken', text: 'Team Token', icon: 'group', iconColor: 'teal',
  },
  {
    value: 'organizationToken', text: 'Organization Token', icon: 'domain', iconColor: 'gray',
  },
  {
    value: 'botToken', text: 'BOT Token', icon: 'memory', iconColor: 'green',
  },
  {
    value: 'webhookKey', text: 'Webhook Key', icon: 'link', iconColor: 'gray',
  },
  {
    value: 'integrationToken', text: 'Integration Token', icon: 'merge_type', iconColor: 'purple',
  },
  {
    value: 'servicePrincipal', text: 'Service Principal', icon: 'verified_user', iconColor: 'yellow',
  },
  {
    value: 'userManagedIdentity', text: 'User Managed Identity', icon: 'supervised_user_circle', iconColor: 'red',
  },
  {
    value: 'cognitoToken', text: 'Cognito Token', icon: 'login', iconColor: 'teal',
  },
  {
    value: 'lambdaAuthorizer', text: 'Lambda Authorizer', icon: 'flash_on', iconColor: 'yellow',
  },
];

export default {
  ACCOUNT_TYPES,
  ACCOUNT_SUBTYPES,
};
