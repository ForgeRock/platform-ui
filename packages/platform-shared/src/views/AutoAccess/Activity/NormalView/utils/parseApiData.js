/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { colorsHex, colorsReadable } from './colors';

export default function parseApiData(response) {
  const categories = {};
  const categoryHasOther = {};
  const categoryNames = [...new Set(response.result.map((item) => item.category))];
  categoryNames.forEach((category) => {
    let categoryValue = {};
    let otherCount = 0;
    let otherCategoryData = {};
    categories[category] = [];
    categoryHasOther[category] = false;

    /**
     * This reorganizing of the API data is needed since the data provided is formatted in a way needed by the UI.
     *
     * The data is flat and need to be organized into a usable object structure.
     * It needs to be manipulated by merging values of 'other', and null together into a single field.
     * The field of 'Others' also need to be converted to have the label 'other' and be sorted last
     */
    response.result.forEach((item) => {
      if (item.category === category) {
        // 'other' or null
        if (item.value === 'other' || !item.value) {
          otherCount += parseInt(item.count, 10);
          if (!categoryHasOther[category]) {
            otherCategoryData = {
              label: 'unknown',
              value: otherCount,
              valueFormatted: otherCount,
            };
            categoryHasOther[category] = true;
          } else {
            otherCategoryData.value = otherCount;
            otherCategoryData.valueFormatted = otherCount;
          }
        } else {
          // Handle all normal categories, except 'Others' which gets renames to 'other'
          categoryValue = {
            label: item.value === 'Others' ? 'other' : item.value,
            value: parseInt(item.count, 10),
            valueFormatted: item.count,
          };
          categories[item.category].push(categoryValue);
        }
      }
    });

    // We have found 'other' or null and need to add the manipulated field to the main array
    if (categoryHasOther[category]) {
      categories[category].push(otherCategoryData);
    }

    // Sort categories based on numerical 'value', but leave 'other' sorted last
    Object.keys(categories).forEach((cat) => {
      categories[cat].sort((a, b) => {
        const aVal = parseInt(a.value, 10);
        const bVal = parseInt(b.value, 10);
        if (a.label === 'other') { return 1; }
        if (b.label === 'other') { return -1; }
        if (aVal < bVal) { return 1; }
        if (aVal > bVal) { return -1; }
        return 0;
      });

      // Set the color values - hex for the Donut Chart, readable for the data table
      let colorIndex = 0;
      categories[cat].forEach((item) => {
        item.color = colorsHex[colorIndex] || colorsHex[4];
        item.colorReadable = colorsReadable[colorIndex] || colorsReadable[4];
        colorIndex += 1;
      });
    });
  });

  return categories;
}
