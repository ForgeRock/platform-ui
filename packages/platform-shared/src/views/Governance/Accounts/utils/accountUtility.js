/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { get } from 'lodash';
import { blankValueIndicator } from '@forgerock/platform-shared/src/utils/governance/constants';
import accountConstants from './accountConstants';

/**
 * Get the badge variant for various account types
 * @returns badge variant
 */
export function getAccountTypeVariant(accountType) {
  const type = accountType.toLowerCase();
  if (type === accountConstants.ACCOUNT_TYPES.DEFAULT) {
    return 'light';
  }
  if (type === accountConstants.ACCOUNT_TYPES.MACHINE) {
    return 'warning';
  }
  return 'light';
}

/**
 * Get the display name from an account
 * @param {object} account Account object
 * @returns Display name
 */
export function getAccountDisplayName(account) {
  return get(account, 'descriptor.idx./account.displayName') || account?.__NAME__ || blankValueIndicator;
}
