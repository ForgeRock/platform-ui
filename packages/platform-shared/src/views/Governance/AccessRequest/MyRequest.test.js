/**
 * Copyright (c) 2023-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount } from '@vue/test-utils';
import * as VueRouter from 'vue-router';
import MyRequest from './MyRequest';
import i18n from '@/i18n';
import { setupTestPinia } from '../../../utils/testPiniaHelpers';
import store from '@/store';

jest.mock('vue-router');

describe('MyRequests', () => {
  function setup(user = { userId: '1234' }) {
    setupTestPinia({ user });
    return mount(MyRequest, {
      global: {
        plugins: [i18n],
        stubs: ['RequestsTab', 'RequestTypesTab'],
      },
    });
  }

  it('should load header texts correctly', () => {
    VueRouter.useRoute.mockReturnValue({
      params: {
        requestsTab: '',
      },
    });
    const wrapper = setup();

    expect(wrapper.find('h1').text()).toBe('pageTitles.MyRequests');
    expect(wrapper.find('p').text()).toBe('pages.myRequests.subTitle');
  });

  it('should load initial state correctly for admin with list tab selected when no tab is setted on the route and governance dev flag is enabled', async () => {
    VueRouter.useRoute.mockReturnValue({
      params: {
        requestsTab: '',
      },
    });
    store.state.SharedStore.governanceDevEnabled = true;

    const wrapper = setup({
      userId: '1234',
      idmRoles: ['openidm-admin'],
    });
    await wrapper.vm.$nextTick();

    const tabsWrapper = wrapper.find('.tabs');
    expect(tabsWrapper.exists()).toBe(true);
    const tabs = wrapper.findAll('[role="tab"]');
    expect(tabs.length).toBe(2);
    expect(tabs[0].text()).toBe('pageTitles.MyRequests');
    expect(tabs[1].text()).toBe('Request Types');
    expect(tabs[0].classes()).toContain('active');
    const requestsTab = wrapper.findComponent({ name: 'RequestsTab' });
    expect(requestsTab.exists()).toBe(true);
  });

  it('should load initial state correctly for admin with list tab selected when no tab is setted on the route and governance dev flag is disabled', async () => {
    VueRouter.useRoute.mockReturnValue({
      params: {
        requestsTab: '',
      },
    });
    store.state.SharedStore.governanceDevEnabled = false;

    const wrapper = setup({
      userId: '1234',
      idmRoles: ['openidm-admin'],
    });
    await wrapper.vm.$nextTick();

    const tabsWrapper = wrapper.find('.tabs');
    expect(tabsWrapper.exists()).toBe(true);
    const tabs = wrapper.findAll('[role="tab"]');
    expect(tabs.length).toBe(1);
    expect(tabs[0].text()).toBe('pageTitles.MyRequests');
    expect(tabs[0].classes()).toContain('active');
    const requestsTab = wrapper.findComponent({ name: 'RequestsTab' });
    expect(requestsTab.exists()).toBe(true);
  });

  it('should load initial state correctly for non-admin with list tab selected when no tab is setted on the route', async () => {
    VueRouter.useRoute.mockReturnValue({
      params: {
        requestsTab: '',
      },
    });
    store.state.SharedStore.governanceDevEnabled = true;

    const wrapper = setup({
      userId: '1234',
      idmRoles: [],
    });
    await wrapper.vm.$nextTick();

    const tabsWrapper = wrapper.find('.tabs');
    expect(tabsWrapper.exists()).toBe(false);
    const requestsTab = wrapper.findComponent({ name: 'RequestsTab' });
    expect(requestsTab.exists()).toBe(true);
  });

  it('should load initial state correctly for admin with Request Types tab selected when tab is setted on the route', async () => {
    VueRouter.useRoute.mockReturnValue({
      params: {
        requestsTab: 'request-types',
      },
    });
    store.state.SharedStore.governanceDevEnabled = true;

    const wrapper = setup({
      userId: '1234',
      idmRoles: ['openidm-admin'],
    });
    await wrapper.vm.$nextTick();
    console.log(wrapper.html());

    const tabsWrapper = wrapper.find('.tabs');
    expect(tabsWrapper.exists()).toBe(true);
    const tabs = wrapper.findAll('[role="tab"]');
    expect(tabs.length).toBe(2);
    expect(tabs[0].text()).toBe('pageTitles.MyRequests');
    expect(tabs[1].text()).toBe('Request Types');
    expect(tabs[1].classes()).toContain('active');
    const requestTypesTab = wrapper.findComponent({ name: 'RequestTypesTab' });
    expect(requestTypesTab.exists()).toBe(true);
  });
});
