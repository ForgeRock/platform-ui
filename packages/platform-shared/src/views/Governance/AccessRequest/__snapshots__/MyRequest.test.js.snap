// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyRequests @Component Tests Navigates to request details page after clicking on "View Details": console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;
